### ISSUE TITLE
Is ServiceClient thread-safe

### ISSUE THREAD

**Issue by zsybupt** [microsoft/PowerPlatform-DataverseServiceClient#253]
Create a ServiceClient is expensive, is it thread-safe? so we can reuse the same client for multi thread

**Comment by MattB-msft** [comment]
Yes and yes, though depending on your use case you may want or o consider using clones of the connection per thread and using the async versions of the calls.

**Comment by zsybupt** [comment]
@MattB-msft Thanks for your reply, let me describe more about my scenario, we now have a asp.net core service, and we are now using this constructor
![image](https://user-images.githubusercontent.com/3711046/159937364-06a12070-0c6d-4904-b417-a424511fb04a.png)
because we need to access different dataverse environments. I have check the blog you mentioned before https://colinvermander.wordpress.com/2020/03/26/cdsserviceclient-with-asp-net-core/
which is useful when the application just needs to access a single environment.

I am now trying to create a factory like this to create ServiceClient for difference instanceuri
```csharp
public class ServiceClientFactory
{
    public ServiceClient Create(Uri instanceUrl, Func<string, Task<string>> tokenProviderFunction)
    {
       return new ServiceClient(instanceUrl,tokenProviderFunction);
    }
}
```
I have some questions here:
1. can we use useUniqueInstance = false in this situation?
2. because we want to make the creation as fast as possible, how can we achive this, is it ok to have a static dictionary here to store the root ServiceClient and return cloned client every time

```csharp
public class ServiceClientFactory
{
    private static ConcurrentDictionary<Uri,ServiceClient> cache = new ConcurrentDictionary<Uri,ServiceClient>();
    public ServiceClient Create(Uri instanceUrl, Func<string, Task<string>> tokenProviderFunction)
    {
       if(cache.TryGetValue(instanceUrl, out ServiceClient root))
      {
          return root.Clone();
      }
       var tempRoot = new ServiceClient(instanceUrl,tokenProviderFunction);
       var newRoot = cache.GetOrAdd(instanceUrl,tempRoot);
       if(newRoot!=tempRoot)
       {
           tempRoot.Dispose();
       }
       return newRoot.Clone();
    }
}
```
3. And we also want to manage the lifetime of the root ServiceClient, because if there is no more access to one environment, we can release the connection of the root Client. If we try to do this, do we need to ensure that all cloned ServiceClient is unused before we dispose the root ServiceClient.
